### PR TITLE
network: add system_info network activator cloud.cfg overrides

### PR DESCRIPTION
**Pull request by blackboxsw**
## Proposed Commit Message

```
Support overriding network activators in cloud.cfg system_info
on disk.

Default cloud-init activators are used either during hot-plug or
for datasources that are detected during init-network stage when
basic networking has already been setup by the OS.a

Activators are discovered in the following priority order and
determined based on the presence of related network tools:

system_info:
  network:
    activators: [eni, netplan, network-manager, networkd]

On some systems where multiple network config tools are installed
it is necessary to override the priority order of detection to ensure
one activator is chosen over the other. This is done by providing
the a custom cloud config snippet in a /etc/cloud/cloud.cfg.d/*cfg.

LP: #1958377
```

## Addtional context
Also related to ubuntu-desktop-installer [LP #1982855](https://bugs.launchpad.net/livecd-rootfs/+bug/1982855)

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

### CODE REVIEW

**Comment by holmanb** [comment]
Further context: https://bugs.launchpad.net/cloud-init/+bug/1982857

**Review by aciba90 — CHANGES_REQUESTED** [review]
Functional-wise LGTM, thanks!

I think we can improve the new added types, as these variables are actually optional:

```diff
diff --git a/cloudinit/net/activators.py b/cloudinit/net/activators.py
index 5acc6266..c781e29f 100644
--- a/cloudinit/net/activators.py
+++ b/cloudinit/net/activators.py
@@ -1,7 +1,7 @@
 # This file is part of cloud-init. See LICENSE file for license information.
 import logging
 from abc import ABC, abstractmethod
-from typing import Dict, Iterable, List, Type
+from typing import Dict, Iterable, List, Optional, Type
 
 from cloudinit import subp, util
 from cloudinit.net.eni import available as eni_available
@@ -97,7 +97,7 @@ class IfUpDownActivator(NetworkActivator):
     # E.g., NetworkManager has a ifupdown plugin that requires the name
     # of a specific connection.
     @staticmethod
-    def available(target: str = None) -> bool:
+    def available(target: Optional[str] = None) -> bool:
         """Return true if ifupdown can be used on this system."""
         return eni_available(target=target)
 
@@ -269,7 +269,7 @@ NAME_TO_ACTIVATOR: Dict[str, Type[NetworkActivator]] = {
 
 
 def search_activator(
-    priority: List[str] = None, target: str = None
+    priority: Optional[List[str]] = None, target: Optional[str] = None
 ) -> List[Type[NetworkActivator]]:
     if priority is None:
         priority = DEFAULT_PRIORITY
```

File: cloudinit/net/activators.py [hunk 1]
```diff
@@ -272,15 +279,19 @@ def search_activator(
...
+    activator_classes = [NAME_TO_ACTIVATOR[name] for name in priority]
+    return [
+        activator_cls
+        for activator_cls in activator_classes
+        if activator_cls.available(target)
+    ]
 
 
 def select_activator(priority=None, target=None) -> Type[NetworkActivator]:
     found = search_activator(priority, target)
```
Comment (holmanb) [inline cloudinit/net/activators.py:291]: With the typing additions to the `search_activators()` parameters, Pyright warns here.

**Review by holmanb — CHANGES_REQUESTED** [review]
I left some inline nitpicks about typing, but I wanted to check if this override behavior should be extended to hotplug as well. I see that NetHandler in hotplug_hook.py calls select_activator() but doesn't use the new override config option. Don't we want the behavior to match?

File: cloudinit/net/activators.py [hunk 2]
```diff
@@ -254,15 +254,22 @@ def bring_down_interface(device_name: str) -> bool:
...
+    "eni": IfUpDownActivator,
+    "netplan": NetplanActivator,
+    "network-manager": NetworkManagerActivator,
+    "networkd": NetworkdActivator,
+}
+
 
 def search_activator(
-    priority=None, target=None
+    priority: List[str] = None, target: str = None
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply cloudinit/net/activators.py]: Good pt, I typed search_activator when I should have typed select_activator. addressed both and dropped defaults on search_activator (and fixed unittests)

**Comment by blackboxsw** [comment]
added hotplug hook integration via surfacing a Distro.network_activator property  so external callers (the hotplug hook) can drive configured network changes without having to initialize their own Init() to call _exract_config("system_info"). They can instead get the configured network_activator from the distro and perform up/down operations on that.

**Comment by blackboxsw** [comment]
@holmanb @aciba90 just validated hotplug upgrade behavior sees and processes the activators configuration from disk in /etc/cloud/cloud.cfg.d/*

```
2022-07-29 15:59:46,070 - activators.py[DEBUG]: Using selected activator: <class 'cloudinit.net.activators.NetplanActivator'> from priority: ['netplan', 'eni']
2022-07-29 16:00:22,501 - __init__.py[DEBUG]: Selected renderer 'netplan' from priority list: ['netplan', 'eni', 'sysconfig']
2022-07-29 16:00:22,823 - activators.py[DEBUG]: Using selected activator: <class 'cloudinit.net.activators.NetplanActivator'> from priority: ['netplan', 'eni']
```
No traces or concerns here.

**Review by aciba90 — APPROVED** [review]
LGTM. Thanks!

**Review by holmanb — APPROVED** [review]
Thanks!

**Pull request merged**